File: CSharp/704-BinarySearch/704BinarySearch.cs
// BINARY SEARCH
// Time Complexity: O(log n) (Analyze Complexity Verified)
//      You're dividing the input set by 2 each time and ruling out those indexes
// Space Complexity: O(1)
//      We just need the start and end index

public class Solution {
    public int Search(int[] nums, int target) {
        int startIndex = 0;
        int endIndex = nums.Length - 1;
        while (startIndex <= endIndex){
            int midIndex = (startIndex + endIndex)/2;
            if (target == nums[midIndex]) return midIndex;
            else if (target < nums[midIndex]) endIndex = midIndex -1;
            else startIndex = midIndex + 1;
        }
        return -1;
    }
}

File: Javascript/704-BinarySearch/704BinarySearch.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
// Time complexity: O(log n). Divide in 2 every time.
// Space Complexity: O(1) just high and low pointers.
var search = function (nums, target){
  let low = 0,
      high = nums.length - 1;
  while (low <= high){
      let mid = Math.floor((low + high) / 2);
      if (nums[mid] === target) return mid;
      else if (target < nums[mid]) high = mid - 1;
      else low = mid + 1;
  }
  return -1;
}